### PR TITLE
Add kapelse readers support in COMPOSITE_AS_MULTISLOT mode

### PR DESCRIPTION
**Pull request by bmoraine**
Hello,

Please find 2 patches regarding COMPOSITE_AS_MULTISLOT mode to simplify its implementation and to add support for Kapelse readers supports.

Best Regards

Baptiste Moraine

### CODE REVIEW

**Comment by bmoraine** [comment]
Build issue when --enable-strict --enable-composite-as-multislot used for configuration is now fixed by last commit

File: src/ccid_usb.c [hunk 1]
```diff
@@ -644,6 +644,39 @@ status_t OpenUSBByName(unsigned int reader_index, /*@null@*/ char *device)
 				}
 #endif
 
+#ifdef USE_COMPOSITE_AS_MULTISLOT
+				if (VENDOR_KAPELSE == GET_VENDOR(readerID))
+				{
+					switch (readerID)
+				 	{
+						case KAPELSE_KAPECV :
```
Comment (LudovicRousseau) [inline src/ccid_usb.c:652]: Why to you want to handle this reader as a multislot reader if it has only 1 CCID interface?

File: src/ccid_usb.c [hunk 2]
```diff
@@ -644,6 +644,39 @@ status_t OpenUSBByName(unsigned int reader_index, /*@null@*/ char *device)
...
+						    /* KAP-eCV : only first interface is a ccid one */
+						    max_interface_number=0;
+						    DEBUG_INFO1("Kapelse reader forced as monoslot!");
+						    break;
+						    
+						case KAPELSE_KAPLIN2 :
+						    /* KAP&LINK2 : only 3 first interfaces are ccid ones */
+						    if(config_desc->bNumInterfaces > 3)
```
Comment (LudovicRousseau) [inline src/ccid_usb.c:660]: Why this test?
In my list the KAP-LINK2 has 3 CCID interfaces https://ccid.apdu.fr/ccid/shouldwork.html#0x29470x0105

Do you have another model with a different configuration but using the same `idProduct`?

**Review by LudovicRousseau — CHANGES_REQUESTED** [review]
I pushed your fist commit with changes in https://github.com/LudovicRousseau/CCID-debug/

Please re-submit your patch in this `-debug` repo.

**Comment by LudovicRousseau** [comment]
Can you also give me the output of the following command for each kapelse readers please
```
lsusb -d 2947: -v | grep 'bInterfaceClass\|idVendor\|idProduct'
```
I need to know what other interfaces are present and in which position.

For example the Gemalto Prox DU reader has 2 CCID interfaces and 1 HID interface. The HID interface is in the first position so `max_num_interfaces` is not the same as the number of "slots" as you suggested in your patch. The code you removed looked like a duplication but was not exactly the same.

File: src/ccid_usb.c (hunk 1 above)
Comment (bmoraine) [reply src/ccid_usb.c:652]: You are right, KAP-eCV, there is no need to handle it as a multislot reader. It is a USB composite with only one CCID interface (the first one), a HID interface and a CDC-ACM interface.

File: src/ccid_usb.c (hunk 2 above)
Comment (bmoraine) [reply src/ccid_usb.c:660]: This test is needed as KAP-LINK2 is a USB composite device presenting either 3 CCID interfaces or 3 CCID interfaces (the 3 first ones) and a CDC-ACM interface according user configuration

Comment (LudovicRousseau) [reply src/ccid_usb.c:660]: I see.
It should work fine if you something like use:
```
max_interface_number = 3; /* up to 4 interfaces */
num_CCID_interfaces = 3;  /* 3 CCID interfaces */
```

**Comment by bmoraine** [comment]
Please find below the output of the command :
`lsusb -d 2947: -v | grep 'bInterfaceClass\|idVendor\|idProduct' `

For Kap&Link2 
- When enumerating 3 CCID interfaces :
`$ lsusb -d 2947: -v | grep 'bInterfaceClass\|idVendor\|idProduct'
  idVendor           0x2947 
  idProduct          0x0105 
      bInterfaceClass        11 Chip/SmartCard
      bInterfaceClass        11 Chip/SmartCard
      bInterfaceClass        11 Chip/SmartCard
`
- When enumerating 3 CCID interfaces and 1 CDC ACM Interface :
`$ lsusb -d 2947: -v | grep 'bInterfaceClass\|idVendor\|idProduct'
  idVendor           0x2947 
  idProduct          0x0105 
      bInterfaceClass        11 Chip/SmartCard
      bInterfaceClass        11 Chip/SmartCard
      bInterfaceClass        11 Chip/SmartCard
      bInterfaceClass         2 Communications
      bInterfaceClass        10 CDC Data
      `
For KAP-eCV :
`$ lsusb -d 2947: -v | grep 'bInterfaceClass\|idVendor\|idProduct'
  idVendor           0x2947 
  idProduct          0x0112 
      bInterfaceClass        11 Chip/SmartCard
      bInterfaceClass         3 Human Interface Device
      bInterfaceClass         2 Communications
      bInterfaceClass        10 CDC Data
`
For all others Kapelse readers there are as many CCID interfaces as USB interfaces as for Kap&Link :
  idVendor           0x2947 
  idProduct          0x0101 
      bInterfaceClass        11 Chip/SmartCard
      bInterfaceClass        11 Chip/SmartCard

**Comment by bmoraine** [comment]
It's unclear for me. Do you want me to submit Kapelse reader suport in COMPOSITE_AS_MULTISLOT mode patch in this -debug repo? If yes, via a pull request?

**Comment by LudovicRousseau** [comment]
I made modifications to your first patch in the [CCID-debug repo](https://github.com/LudovicRousseau/CCID-debug/commit/951739871d55326f66af32837cdcc31d45898b2a).
You now have to define 2 values: `max_interface_number` and `num_CCID_interfaces`.

Since the CCID interfaces on the kapelse readers are in the first interfaces `num_CCID_interfaces` should contain the number of CCID interfaces (easy) and `max_interface_number` should be `num_CCID_interfaces -1` because it starts at 0.

Yes, I ask you to propose a Pull Request on this repo to add support of kapelse readers.
I do not have any of the kapelse readers so I can't test the code myself. 
I don't want to publish untested code.

**Comment by bmoraine** [comment]
For information, I have created a Pull Request as recommended : 
https://github.com/LudovicRousseau/CCID-debug/pull/1

**Comment by LudovicRousseau** [comment]
Good.
have you checked your patches works fine on macOS with your readers?

**Comment by bmoraine** [comment]
No, I tested only on linux using "./configure --enable-strict --enable-composite-as-multislot "

**Comment by LudovicRousseau** [comment]
I was expecting an answer like that :-)
I asked because your patch does not correctly set (I think) the `max_interface_number` value.
pcsc-lite on GNU/Linux **does** support composite devices. So you can't check your code on Linux.

You have to use macOS to test your changes work fine on macOS.

Please also test connecting 2 Kapelse readers and check all the CCID interfaces are visible using pcsctest.
The CCID driver has code to reset `static_interface` value when all the CCID interfaces have been used, so the driver is ready to connect another "composite-as-multislot" reader.
https://github.com/LudovicRousseau/CCID/blob/master/src/ccid_usb.c#L732

**Comment by LudovicRousseau** [comment]
@bmoraine you can also send me Kapelse sample readers so I can make tests on my side.
Contact me by private email.

**Comment by bmoraine** [comment]
@LudovicRousseau I will test Kapelse readers on macOS Sonoma and ,in the same time, send you some readers.

Regarding ccid driver generatioin on macos, I always slightly modified ccid driver source in SmartcardCCID :
https://github.com/apple-oss-distributions/SmartcardCCID/ and never update the entire ccid driver source files.

In my undestanding, Apple applies patches on top of cdid driver and use specific parameters when configuring ccid project according  :
https://github.com/apple-oss-distributions/SmartcardCCID/blob/main/ccid/Makefile.
I assume that if I replace ccid tar.gz with the content of CCID-debug and I execute the makefile, it should be OK. Do you confirm?

**Comment by LudovicRousseau** [comment]
> I assume that if I replace ccid tar.gz with the content of CCID-debug and I execute the makefile, it should be OK. Do you confirm?

I also assume it should work.
I never used the Apple process to build my CCID driver on macOS, so you are the expert on this point.

**Comment by bmoraine** [comment]
What is your process to build CCID driver on macOS?

**Comment by LudovicRousseau** [comment]
https://ccid.apdu.fr/#MacOSX

File: src/ccid_usb.c (hunk 2 above)
Comment (LudovicRousseau) [reply src/ccid_usb.c:660]: @bmoraine Is it possible for the KAP-LINK2 to NOT have 3 CCID interfaces?

Why can't we use ``max_interface_number = 2;`` in all cases without checking the value of ``config_desc->bNumInterfaces``?

Comment (bmoraine) [reply src/ccid_usb.c:660]: @LudovicRousseau Effectively, KAP-LINK2 has always 3 CCID interfaces (the 3 first ones) and an optional 4th CDC-ACM interface. So, you are right, we can use max_interface_number = 2; in all cases without checking the value of config_desc->bNumInterfaces

**Comment by LudovicRousseau** [comment]
Fixed in 73d9d3a9bd9e2905912a76d7e4615a9729d83fc2 and later patches

**Pull request closed without merging**